### PR TITLE
[WIP] Changed LexBuffer type from char to uint16

### PR DESCRIPTION
**Pull request by ncave**
Rationale:

Changing the LexBuffer type from char to uint16 has very little or no performance impact on FCS on .NET, but provides a substantial performance boost on platforms that don't implement chars as uint16 (e.g. Fable).

```
Build Target: Test

----------------------------------------------------------
Build Time Report (Before this PR):
---------------------------------------------------------------------
Target     Duration
------     --------
Start      00:00:00.0003701
Restore    00:00:02.9641905
Build      00:01:56.1463016
Test       00:07:54.6751388
Total:     00:09:53.8885789
---------------------------------------------------------------------
Status:    Ok
---------------------------------------------------------------------

---------------------------------------------------------------------
Build Time Report (After this PR):
---------------------------------------------------------------------
Target     Duration
------     --------
Start      00:00:00.0001706
Restore    00:00:02.9515601
Build      00:01:56.0570723
Test       00:07:53.6562371
Total:     00:09:52.7613661
---------------------------------------------------------------------
Status:    Ok
---------------------------------------------------------------------
```

### CODE REVIEW

**Comment by TIHan** [comment]
I don't know much about Fable; why does Fable not represent `char` as a `uint16` ?

**Comment by TIHan** [comment]
I'm really ignorant on Fable, and don't really know how it does the mapping. Can someone explain this a bit more?

**Comment by TIHan** [comment]
Also, can we get performance numbers on Fable as well?

File: src/utils/prim-lexing.fs [hunk 1]
```diff
@@ -167,17 +166,17 @@ namespace Internal.Utilities.Text.Lexing
             LexBuffer<'Char>.FromArrayNoCopy buffer
 
         // Important: This method takes ownership of the array
-        static member FromChars (arr:char[]) = LexBuffer.FromArrayNoCopy arr
```
Comment (TIHan) [inline src/utils/prim-lexing.fs]: We are now making a new array. Before, the intent was not to do so.

File: src/utils/prim-lexing.fs [hunk 2]
```diff
@@ -167,17 +166,17 @@ namespace Internal.Utilities.Text.Lexing
             LexBuffer<'Char>.FromArrayNoCopy buffer
 
         // Important: This method takes ownership of the array
-        static member FromChars (arr:char[]) = LexBuffer.FromArrayNoCopy arr 
+        static member FromChars (arr:char[]) = LexBuffer.FromArrayNoCopy (arr |> Array.map uint16)
```
Comment (TIHan) [inline src/utils/prim-lexing.fs]: We are now making a new array. Before, the intent was not to do so.

Comment (ncave) [reply src/utils/prim-lexing.fs]: Yes, looks like this one can be improved, as it's always used with a new array, like that: `LexBuffer<char>.FromChars (s.ToCharArray())`
I'll fix it.

File: src/absil/illex.fsl [hunk 3]
```diff
@@ -16,7 +16,7 @@ open Microsoft.FSharp.Compiler.AbstractIL.Internal.AsciiParser
 open Microsoft.FSharp.Compiler.AbstractIL.Internal.AsciiConstants 
 
 
-let lexeme (lexbuf : LexBuffer<char>) = new System.String(lexbuf.Lexeme)
+let lexeme (lexbuf : LexBuffer<uint16>) = new System.String(lexbuf.Lexeme |> Array.map char)
```
[the remark this answers is not in the document]
Comment (ncave) [reply src/absil/illex.fsl]: Thanks, I'll look at the usage pattern for possible improvements, as sometimes only a single char from the resulting string is retrieved.

**Comment by ncave** [comment]
@TIHan The char type in Fable is currently implemented as a string. Although technically it is possible to implement it as uint16, it was decided against doing that, perhaps @alfonsogarciacaro can elaborate more. The performance boost can be [quite large](https://github.com/fable-compiler/Fable/issues/1457#issuecomment-402584762).

**Comment by alfonsogarciacaro** [comment]
There's no char in JS, so every time you split/enumerate/index a string, you get another string of length 1. We tried to hijack these actions and convert the results to numbers so we could have a more faithful representation of F# `char` (and get better performance in some situations as @ncave points). However, in situations where we don't have type information at compile time (generic seq, boxed arguments...) this was tricky and we got unexpected results. That's the reason why we opted to match JS native behavior and use strings to represent `char`.

**Comment by KevinRansom** [comment]
@ncave can you teach fable to treat, LexBuffer<char> as if it were LexBuffer<uint16>?  

I don't think it would be a good precedent for us to turn FCS APIs that use char[] and string into uint16[] merely because it runs faster in javascript.

As @TIHan pointed out it would ruin debuggability for us because the debugger would represent each char as a hex number.


Kevin

**Comment by forki** [comment]
@kevinransom fable has become one of the most popular use cases of F#. Making it faster can be a overall win for F#. Maybe we can add a debugger view to mitigate the pain? Or conditional build for debugging the Lexer? I know that's not super nice, but maybe a good compromise?

**Comment by ncave** [comment]
Just to clarify, it's not Fable that gets performance benefit from this PR, it's the FCS transpiled with Fable as part of Fable REPL. This PR is just trying to bring upstream parts of the Fable REPL PR, but it's fine if it needs to stay downstream for the time being. Please feel free to close it at any time, or I can close it after I finish applying the review suggestions above.

**Comment by alfonsogarciacaro** [comment]
It's not ideal but maybe making some type alias for Fable?

```fsharp
#if FABLE_COMPILER
type LexBuffer = LexBuffer<uint16>
#else
type LexBuffer = LexBuffer<char>
#endif
```

**Comment by KevinRansom** [comment]
@alfonsogarciacaro if it works for Fable, then thats great.

**Comment by KevinRansom** [comment]
@alfonsogarciacaro,

whilst I am okay with a simple #if to address your problem this proposal is getting push back here from others that "cluttering" up our source code with #if's for compilers we don't build is not such a great choice.

The consensus is that you should "fix up" your codegen stage so that it maps arrays of chars to be performant or you provide out of band metadata to your back end that tells it that:

````
LexBuffer<char> should be treated like LexBuffer<int16>
````

See what I did there? I threw unnamed **others** under the bus.

Kevin

**Comment by forki** [comment]
@KevinRansom "fix up your codegen stage"? What exactly do you mean? @ncave is transpiling the FCS to js - then the FCS runs as a normal javascript program in the browser. ou mean we should change the fable transpiler so that it never transpiles `LexBuffer<char>` as such? I guess @alfonsogarciacaro can make that happen, but then Fable would need to learn about that special internal type.

**Comment by ncave** [comment]
@TIHan Thanks for reviewing, I appreciate it. Most of the extra allocations were [removed](https://github.com/Microsoft/visualfsharp/pull/5614/commits/b095a1bdd9738189aaa2ac2d349e297a3232ec26), but the sentiment seems to be against creating conditional LexBuffer types here, so it's probably better to close this PR and keep it downstream for now.

**Comment by KevinRansom** [comment]
@forki snipped from above:

The char type in Fable is currently implemented as a string. Although technically it is possible to implement it as uint16, it was decided against doing that, perhaps @alfonsogarciacaro can elaborate more. The performance boost can be quite large.

When transpiling LexBuffer<char>  instead of transpiling the char as string, transpile is int16.  Ideally add some metadata with a table of types that can gain this optimisation so that we don't need to continue to convert T<char>s to T<int16> as you discover other interesting types that can benefit from this optimisation.

This would effectively be the same thing as us changing our source code, except, fable has a specific T<char> optimisation built in.

**Comment by forki** [comment]
Javascript has int16 type? AFAIK it only has number.

Kevin Ransom (msft) <notifications@github.com> schrieb am Mi., 12. Sep.
2018, 10:38:

**Comment by KevinRansom** [comment]
Then treat it as array<number> whatever you would map <uint16> to.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by alfonsogarciacaro** [comment]
Thanks for your suggestion Kevin. Replacing a full type in Fable is unfortunately not so easy though. Besides the fact that Fable 2 keeps a reference to FSharpEntity for most types and I don't know how to create a new FSharpEntity, as far as I can see in the PR most of the functions dealing with the char array are external to LexBuffer so we would have to hijack all of this.

Fable has a replacement system used to support BCL and FSharp.Core, but it occurs at the call level and is only meant for stable APIs. If we did it here, it would silently break after any change to related code in the F# compiler.

Anyways I understand it's difficult to think of different platforms in the source code of the compiler. It'd be a good opportunity to increase confidence in the future of Fable to see that it has some capacity of influencing the F# compiler though :) 

> BTW, JS doesn't have uint16 but it does have typed numeric arrays, that's why we get a performance boost when changing from char[] to uint16[].

**Comment by KevinRansom** [comment]
@alfonsogarciacaro 

Fable is an important scenario to us. Can you measure and report the performance boost of this change in the fable compiler?

As I said, I am okay with a #if approach, I will push back on the push back; however, I do think you should consider what it would take to optimise the code gen, but I understand that stuff is subtle and comes with all sorts of unanticipated effects.

So If you believe the performance boost to be significant let's proceed.  I think you should probably also figure out how to get an acceptance test case that we can run in our CI along the lines of the fcs suite to make sure we don't break you.

Kevin

**Comment by ncave** [comment]
@KevinRansom To clarify, as stated above, there is no performance impact on the Fable compiler itself (as it runs on .NET core). The performance boost is only for FCS library (when transpiled with Fable to javascript), which is used by the Fable REPL (technically not part of Fable, just a showcase that Fable can compile itself).

**Comment by KevinRansom** [comment]
@ncave 

so ... 
- Does the Fable repl perform noticably faster with this change? 
- Have you measured the change, what was the improvement?
- Does any other tool or product use FCS and can they benefit from the change?


Kevin

**Comment by ncave** [comment]
@KevinRansom Yes, the performance difference is [substantial](https://github.com/fable-compiler/Fable/issues/1457#issuecomment-402584762). Technically the Fable REPL is bundled as a component so it can potentially be used on other websites, or with Node.js.

IMO, the arguments you raised are valid, and they were already [discussed](https://github.com/fable-compiler/Fable/issues/1524) and [considered](https://github.com/fable-compiler/Fable/pull/1473) at length in the Fable community. Implementations already exist for char represented as both `string` and `uint16`, but it was ultimately decided to stick with string representation for better interoperability with JavaScript.

I apologize if this PR is stirring unnecessary controversy, that was not my intent at all. It's perfectly fine for this PR to stay downstream, it's not like the [FCS Fable REPL](https://github.com/fsharp/FSharp.Compiler.Service/pull/688) PR is going anywhere anytime soon.

**Comment by KevinRansom** [comment]
@ncave, hardly a controversy mate, I am glad you have a solution you can use.

Thanks

Kevin

**Pull request closed without merging**